src/Expression/Differentiator.purs

```diff
@@ -0,0 +1,100 @@
+module Expression.Differentiator (differentiate, secondDifferentiate) where
+
+import Prelude
+import Expression.Syntax (BinaryOperation(..), Expression(..), UnaryOperation(..))
+
+secondDifferentiate :: Expression -> Expression
+secondDifferentiate = differentiate <<< differentiate
+
+differentiate :: Expression -> Expression
+differentiate = case _ of
+  ExpressionLiteral _ -> ExpressionLiteral 0.0
+  ExpressionVariable _ -> ExpressionLiteral 1.0
+  ExpressionBinary operation leftExpression rightExpression -> differentiateBinaryOperation operation leftExpression rightExpression
+  ExpressionUnary operation expression -> differentiateUnaryOperation operation expression
+
+differentiateBinaryOperation :: BinaryOperation -> Expression -> Expression -> Expression
+differentiateBinaryOperation (Plus) leftExpression rightExpression = ExpressionBinary Plus (differentiate leftExpression) (differentiate rightExpression)
+
+differentiateBinaryOperation (Minus) leftExpression rightExpression = ExpressionBinary Plus (differentiate leftExpression) (differentiate rightExpression)
+
+differentiateBinaryOperation (Times) leftExpression rightExpression = ExpressionBinary Plus (ExpressionBinary Times u v') (ExpressionBinary Times u' v)
+  -- Product rule
+  where
+  u = leftExpression
+
+  v = rightExpression
+
+  u' = differentiate u
+
+  v' = differentiate v
+
+differentiateBinaryOperation (Divide) topExpression bottomExpression = ExpressionBinary Divide (ExpressionBinary Minus (ExpressionBinary Times f' g) (ExpressionBinary Times f g')) (ExpressionBinary Power g (ExpressionLiteral 2.0))
+  -- Quotient rule
+  where
+  f = topExpression
+
+  g = bottomExpression
+
+  f' = differentiate f
+
+  g' = differentiate g
+
+differentiateBinaryOperation (Power) (ExpressionVariable "e") (ExpressionLiteral value) = ExpressionUnary Exp (ExpressionLiteral value)
+
+differentiateBinaryOperation (Power) (ExpressionVariable "e") exponentExpression = ExpressionBinary Times (differentiate exponentExpression) (ExpressionUnary Exp exponentExpression)
+
+differentiateBinaryOperation (Power) mantisaExpression (ExpressionLiteral value) = ExpressionBinary Times (ExpressionLiteral value) (ExpressionBinary Power mantisaExpression (ExpressionLiteral (value - 1.0)))
```

Comment:
Please rename mantisaExpression to baseExpression.  Mantissa means something a bit different as used eg in floating-point number representations.

Reply:
Resolved in the latest commit